### PR TITLE
[DDO-2166] Sign and notarize macOS releases + extras!

### PR DESCRIPTION
**Pull request by em-may**
This PR adds signing and notarizing for macOS builds. It also adds the following goodies:
* Go deps are now cached between builds
* Builds are parellelized
* The docker build uses the linux release artifact instead of doing another build

Overall this speeds up builds a lot and minimizes the time impact that adding signing/notarizing would otherwise have.

Some notes on time improvements/speed ups, using [this](https://github.com/broadinstitute/thelma/runs/7362013023?check_suite_focus=true) before example and [this](https://github.com/broadinstitute/thelma/actions/runs/2707986760) after example:
* Build times
  * Before: 5m 53s
  * **After: 1m59s**
* Signing and notarizing:
  * Before: N/A
  * After: 4m 35s
* Image gen + push + release push:
  * Before: 3m 43s
  * **After: 1m 47s**
* Total build GHA run times:
  * Before: 10m 0s
  * After: 10m 0s

TL;DR: Signing and notarizing for macOS releases now works and it adds basically no time to the build workflow despite adding a step that takes 4.5mins.

### CODE REVIEW

File: .github/actions/make/action.yaml [hunk 1]
```diff
@@ -22,12 +22,16 @@ inputs:
...
-    - OUTPUT_DIR=${{ inputs.output-dir }}
-    - OS=${{ inputs.os }}
-    - ARCH=${{ inputs.arch }}
-    - VERSION=${{ inputs.version }}
+  using: 'composite'
+  steps:
+    # Cache go deps across builds
+    - uses: actions/setup-go@v3
+      with:
+        go-version: '1.17'
```
Comment (choover-broad) [inline .github/actions/make/action.yaml:30]: do we need a `cache: true` here to actually cache dependencies?

File: .github/workflows/build.yaml [hunk 2]
```diff
@@ -35,15 +35,32 @@ env:
...
     runs-on: ubuntu-latest
+    outputs:
+      version: ${{ steps.tag.outputs.tag }}
     steps:
       - name: Checkout current code
         uses: actions/checkout@v2
         with:
           token: ${{ secrets.BROADBOT_TOKEN }}
 
+      - name: Cache build directory
```
Comment (choover-broad) [inline .github/workflows/build.yaml:55]: so just checking my understanding, this isn't actually caching as much as "save contents of build output directory and make available for the notarization job"?

File: .github/workflows/build.yaml [hunk 3]
```diff
@@ -56,33 +73,100 @@ jobs:
...
+      - name: Pull cache
+        id: pull-cache
+        uses: actions/cache@v3
+        env:
+          cache-name: build-dir-cache
+        with:
+          path: ./output
+          key: commit-${{ github.sha }}-job-${{ github.run_id }}
+
+      # A second cache is needed to force GH to upload the new macOS tarballs,
```
Comment (choover-broad) [inline .github/workflows/build.yaml:117]: maybe call the two "caches" `unsigned-build-artifacts` and `signed-build-artifacts` for clarity? IDK, it's a matter of taste but that would have helped me personally understand sooner 🤷  (Since caching is not what we're doing as much as copying artifacts across jobs)

File: Makefile [hunk 4]
```diff
@@ -39,6 +39,21 @@ else
 	RUNTIME_DEPS_TESTEXEC=true
 endif
 
+# MACOS_SIGN_AND_NOTARIZE: When false, creates a release tarball as part of the make process,
```
Comment (choover-broad) [inline Makefile:42]: One thing that's confusing to me about these changes is that even though the Makefile has a MACOS_SIGN_AND_NOTARIZE variable, if I set it to true, my binaries are not going to be notarized :) All of that happens elsewhere, in GHA pipelines, right?

I almost think it might be more clear to not change this file... We can leave it as-is, right? It's not like producing the tarball deletes the unsigned binaries from the cache...

File: Dockerfile [hunk 5]
```diff
@@ -1,30 +1,23 @@
...
-
 #
 # Copy dist into runtime image
 #
-FROM alpine:${ALPINE_IMAGE_VERSION} as runtime
+FROM alpine:${ALPINE_IMAGE_VERSION}
+
+ARG THELMA_LINUX_RELEASE
+
+COPY output/releases/${THELMA_LINUX_RELEASE} .
```
Comment (choover-broad) [inline Dockerfile:10]: Just want to say that this alone is a fantastic speedup! So great to see the binaries re-used instead of another compile! Thank you!

File: .github/workflows/build.yaml [hunk 6]
```diff
@@ -56,33 +73,100 @@ jobs:
...
+          security create-keychain -p temp-kc-pwd ./output/kc/release.keychain
+          security default-keychain -s ./output/kc/release.keychain
+          security unlock-keychain -p temp-kc-pwd ./output/kc/release.keychain
+          security import ./output/kc/certificate.p12 -k ./output/kc/release.keychain -P ${THELMA_MACOS_CERT_PWD} -T /usr/bin/codesign
+          security set-key-partition-list -S apple-tool:,apple:,codesign: -s -k temp-kc-pwd ./output/kc/release.keychain
+
+      - name: Sign and notarize macOS releases and create tarball
+        id: san-tarball
+        run: |
+          ./scripts/sign-and-notarize.sh ./output ./output/releases/thelma_${{ needs.build.outputs.version }}_darwin_amd64.tar.gz
```
Comment (choover-broad) [inline .github/workflows/build.yaml:144]: ~~I have a very stupid question. Does output here contain the `arm64` or `amd64` thelma? Is it just whatever was most recently compiled? I don't understand how `sign-and-notarize` could be run in succession on the same output directory and produce results for different platforms :confused:~~

File: Makefile (hunk 4 above)
Comment (choover-broad) [reply Makefile:42]: OK actually I think this variable might never be set to true, since it looks like `sign-and-notarize.sh` unpacks and repacks the tarball. So it seems like we could definitely leave as-is.

File: .github/workflows/build.yaml (hunk 6 above)
Comment (choover-broad) [reply .github/workflows/build.yaml:144]: Okay I was confused, it looks like the ./output parameter here is just a tmp directory and not input. The 2nd param is the name of the tarball to replace (so input and output)

**Review by choover-broad — APPROVED** [review]
Looks good! Great to see such a huge bonus perf improvement in thelma's build process!

File: .github/actions/make/action.yaml (hunk 1 above)
Comment (em-may) [reply .github/actions/make/action.yaml:34]: Nope! Turns out if you do that it will cache between workflows, which we don't want. By default, it uses `actions/cache` (I think) under the hood to give the behavior we want (i.e. caching between jobs within the workflow). The proof is that the build jobs no longer have output showing that they are pulling Go deps.

File: .github/workflows/build.yaml (hunk 3 above)
Comment (em-may) [reply .github/workflows/build.yaml:270]: In the time since this comment, all of the intermediary caches have been given platform-specific names and only the final release cache is called `release-cache`. I've also added a big comment to the top of this file explaining how the caches work for future reference. I think that should do it.

File: scripts/sign-and-notarize.sh [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+
+	# Go back to the previous directory
+	cd - > /dev/null
+
+	# Other fns get the signed zip file from stdout
+	echo "${_outfile}"
+}
+
+# Upload the zip file for notarization and wait for a response
+notarize() {
```
[the remark this answers is not in the document]
Comment (em-may) [reply scripts/sign-and-notarize.sh:82]: Apple notarizes all the files that we upload in the zip, and since notarization requires uploading a file to Apple and waiting for their approval, I think it's best to do it all at once. TL;DR: No, we do not need to notarize each file individually.

File: scripts/sign-and-notarize.sh [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+
+	echo -n "Checking notarization status for ${_sub_id}"
+
+	local _resp=""
+	local _cont=0
+	local _wait_total=0
+	local _sleep_inc=15
+	while
+		# Query Apple to see the status of the notarization request
+		_resp=$(xcrun notarytool log -f json ${CMD_AUTH_FLAGS} ${_sub_id} 2>&1)
```
[the remark this answers is not in the document]
Comment (em-may) [reply scripts/sign-and-notarize.sh:105]: If you look further down you'll see some logic for determining if the response was good or bad. If the submission succeeded, then it's not printed, if it failed, it prints. Initially, I mostly did this to make the output cleaner, and it works, so unless there's a strong desire to change it I'm inclined to leave it since it works.

File: .github/workflows/build.yaml (hunk 6 above)
Comment (em-may) [reply .github/workflows/build.yaml]: Correct, the first param is the output dir, so that a working dir for the script can be created inside it - the second and third are input and output tarballs.
